### PR TITLE
OCR

### PR DESCRIPTION
**Pull request by dkadochnikov**
OCR head with temporary regular expressions (yet only works with lowercase cyrillic and numbers)

### CODE REVIEW

File: src/models/vtr/classifier.py [hunk 1]
```diff
@@ -62,9 +69,11 @@ def __init__(
         self.encoder = nn.TransformerEncoder(encoder_layer, num_layers=num_layers)
         self.norm = nn.LayerNorm(hidden_size)
         self.classifier = nn.Linear(hidden_size, 1)
+        self.ocr = BiLSTM(input_size=256, hidden_size=256, num_layers=2, num_classes=60)
```
Comment (SpirinEgor) [inline src/models/vtr/classifier.py:72]: Let's not hard-code all these numbers, add them to class parameters in any form so in the future we can easily change them.

File: src/models/vtr/embedder.py [hunk 2]
```diff
@@ -123,9 +123,8 @@ def forward(self, slices):
         conv = self.slice_conv(slices.view(batch_size * slice_count, 1, height, width))
 
         _, channels_count, h_out, w_out = conv.shape
-
         batched_conv = conv.view(batch_size, slice_count, channels_count * h_out * w_out)
-        return self.linear_bridge(batched_conv)  # [batch size, slice count, emb size]
+        return self.linear_bridge(batched_conv), conv  # [batch size, slice count, emb size]
```
Comment (SpirinEgor) [inline src/models/vtr/embedder.py:127]: Add to the comment shape of `conv`, please

File: src/models/vtr/ocr.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+from torch import nn
+
+
+class BiLSTM(nn.Module):
```
Comment (SpirinEgor) [inline src/models/vtr/ocr.py:4]: Let's call this `OCRHead` to not confuse with regular LSTM layer

File: src/utils/train.py [hunk 4]
```diff
@@ -147,7 +184,12 @@ def evaluate_model(
 
     precision, recall, f1_score, _ = precision_recall_fscore_support(ground_truth, predictions, average="binary")
     accuracy = accuracy_score(ground_truth, predictions)
-    result = {"accuracy": accuracy, "precision": precision, "recall": recall, "f1": f1_score}
```
Comment (SpirinEgor) [inline src/utils/train.py:150]: This change is unnecessary 🙃

**Review by SpirinEgor — CHANGES_REQUESTED** [review]
There are a lot of changes based on `if ocr_flag`. And most of them because you change what model returns.
I suggest changing the model output to the dict `{"logits": result}`. And in the case of OCR add in it `"ctc_loss": ctc_loss`. Thus, in most cases, you will use logits. And only need to add this small `if`:
```python
if ocr_flag:
    log_dict["bce_loss"] = loss
    log_dict["ctc_loss"] = model_output["ctc_loss"]
    loss = loss + alpha * model_output["ctc_loss"]
log_dict["loss"] = loss
```

Also, I don't like that your model also computes ctc loss. CE computes in one place and CTC in another. What if `VTRClassifier` would return visual embeddings along with logits? And in the train loop, you will use them to compute CTC loss.
```python
bce_loss = criterion(model_output["logits"], batch["labels"])
ctc_loss = compute_ctc_loss(model_output["vtr"], batch["texts"])
```

File: src/datasets/vtr_dataset.py [hunk 5]
```diff
@@ -70,7 +70,13 @@ def __init__(
         ratio: float = 0.7,
     ):
         logger.info(f"Initializing VTRDatasetOCR with {len(labeled_texts)} samples, use max seq len {max_seq_len}")
-        self.labeled_texts = labeled_texts
+
+        self.labeled_texts = []
+        self.char_set: set = set()
+        for sample in labeled_texts:
+            self.labeled_texts.append({"text": clean_text(sample["text"]), "label": sample["label"]})
```
Comment (SpirinEgor) [inline src/datasets/vtr_dataset.py:77]: I suggest checking if the cleaned text is empty and skipping such examples. At the end of init, you can log how many samples you got from the data.

File: src/datasets/vtr_dataset.py [hunk 6]
```diff
@@ -70,7 +70,13 @@ def __init__(
...
+            self.char_set = set.union(self.char_set, set(self.labeled_texts[-1]["text"]))
```
Comment (SpirinEgor) [inline src/datasets/vtr_dataset.py:78]:
```python
self.char_set |= some_string
```

File: src/datasets/vtr_dataset.py (hunk 5 above)
Comment (SpirinEgor) [reply src/datasets/vtr_dataset.py:77]: Also, I think using 2 lists for texts and labels would be more convenient for future using

File: src/models/vtr/classifier.py [hunk 7]
```diff
@@ -74,5 +87,12 @@ def forward(self, input_batch: dict[str, torch.Tensor]) -> torch.Tensor:
...
         encoder_output = encoder_output.mean(dim=1)  # batch_size, emb_size
         encoder_output = self.norm(encoder_output)  # batch_size, emb_size
-        result = self.classifier(encoder_output).squeeze(1)  # batch_size
+        logits = self.classifier(encoder_output).squeeze(1)  # batch_size
+
+        result = {"logits": logits}
+
+        # OCR
+        if self.ocr_flag:
+            result["ocr logits"] = self.ocr(conv)
```
Comment (SpirinEgor) [inline src/models/vtr/classifier.py:96]: It is not good to use spaces in such keys, it may affect in the future somewhere. Just replace them with `_`.

File: src/utils/common.py [hunk 8]
```diff
@@ -22,6 +22,30 @@
...
+    "\u200d"
+    "\u23cf"
+    "\u23e9"
+    "\u231a"
+    "\ufe0f"  # dingbats
+    "\u3030"
+    "]+",
+    re.UNICODE,
+)
+_REG = re.compile("[^а-я0-9 ]", flags=re.MULTILINE)
```
Comment (SpirinEgor) [inline src/utils/common.py:48]: Do we still need this?

File: src/datasets/vtr_dataset.py [hunk 9]
```diff
@@ -70,25 +70,35 @@ def __init__(
         ratio: float = 0.7,
     ):
         logger.info(f"Initializing VTRDatasetOCR with {len(labeled_texts)} samples, use max seq len {max_seq_len}")
-        self.labeled_texts = labeled_texts
+
+        self.texts = []
+        self.labels = []
+        self.char_set: set = set()
+        for sample in labeled_texts:
```
Comment (SpirinEgor) [inline src/datasets/vtr_dataset.py:77]: How fast is this operation? We can add `tqdm` here for better UX 😅

**Review by SpirinEgor — COMMENTED** [review]
Nice work, thank you for this implementation! Just resolve merging conflicts and maybe add progress bar in dataset init and work is done ✅

File: src/utils/common.py [hunk 10]
```diff
@@ -111,6 +113,13 @@ def _set_seed(seed: int):
         torch.cuda.manual_seed_all(seed)
 
 
+def char2int(text: list, char_set: set):
+    char2int_dict = {char: i + 1 for i, char in enumerate(char_set)}
+
+    targets = torch.LongTensor([char2int_dict[c] for c in text])
```
Comment (borisshapa) [inline src/utils/common.py:119]: Blank label must be 0, so you add 1 to all char ids.
But if you have blank symbol in the char set, there'll be two labels for blank symbol.

File: src/utils/train.py [hunk 11]
```diff
@@ -10,13 +10,35 @@
...
 from src.utils.common import set_deterministic_mode, dict_to_device
 from src.utils.config import TrainingConfig
+from src.utils.common import char2int
+from src.datasets.vtr_dataset import VTRDatasetOCR
+
 
 WANDB_PROJECT_NAME = "visual-text"
 
 
+def compute_ctc_loss(criterion: torch.nn.modules.loss.CTCLoss, logits: torch.Tensor, texts: list, char_set: set):
```
Comment (borisshapa) [inline src/utils/train.py:26]: After the order of the dimensions above changes, they will need to be changed in this function as well. Keep in mind that the loss has no option `batch_first`. https://pytorch.org/docs/stable/generated/torch.nn.CTCLoss.html

The loss function takes a tensor of size (T, N, C) or (T, C) where T = sequence length, N = batch size, C = num classes.

Double check that the prediction tensors are correctly related to the ground truth texts

**Review by borisshapa — COMMENTED** [review]
Good job! Clean and understandable code!

**Review by SpirinEgor — APPROVED** [review]
